Fix AES-CTR nonce reusage

textshell opened the pull request:
Questions:
- [x] What to do about local locking?
- [x] What to do about repository side locking? Just depend on #1371? -> yes.
- [x] If so, add assert that repo is locked
- [x] What space reservation should be used?
- [x] Should the reservation be reduced on final commit to avoid nonce space wastage? -> No don’t worry about this
- [x] testing
- [x] diagnostic tools?
- [x] Remove commit with debugging hacks
- [x] What is needed with regard to leaking exceptions from file io?

    Reserve nonce space for AES-CTR before using it.
    
    Reusing the nonce totally breaks AES-CTR confidentiality.
    
    This code uses a reservation of nonce space and stores the next nonce available for
    a future reservation on the client and in the repository.
    
    Local storage is needed to protect about evil repositories that try to gain access to
    encrypted data by not saving nonce reservations and aborting the connection or otherwise
    forcing a rollback.
    
    Storage in the repository is needed to protect against another client writing to the
    repository after a transaction was aborted and thus not seeing the last used nonce from
    the manifest.

    With a real counter mode cipher protection for the multiple client case with an actively 
    evil repository is not possible. But this still protects against cases where the attacker can
    not arbitrarily change the repository but can read everything stored and abort connections
    or crash the server.

    Fixes #22

ThomasWaldmann commented on src/borg/key.py:
```diff
@@ -154,13 +156,82 @@ class AESKeyBase(KeyBase):
+            self.save_repo_end_of_nonce_reservation = lambda next_unreserved, start_nonce: None
+            return None
+        if repo_free_nonce is not None:
+            repo_free_nonce = bytes_to_long(repo_free_nonce)
```
why not directly return the binary nonce from the repo method? would correspond better to get_local_free_nonce also.

textshell replied:
Mostly because i didn’t want to import crypto into the repo class. And borg for some reason seems to not use int.from_bytes. If you prefer i can use int.from_bytes in repository.py and keep an equal interface. I think we still want the wrapper in AESKeyBase/NonceManager for catching InvalidRPCMethod.

textshell replied:
I prefer to use int, as all used need to do arithmetic and having the call sites do the conversion while handling None is messy. Unless you see some problem with sending uint64 sized values as int via msgpack.

ThomasWaldmann replied:
there was some strangeness with msgpack and long ints, have to check how long...

ThomasWaldmann commented on src/borg/key.py:
```diff
@@ -154,13 +156,82 @@ class AESKeyBase(KeyBase):
+            self.commit_repo_nonce_reservation(reservation_end, repo_free_nonce)
+            self.commit_local_nonce_reservation(reservation_end, local_free_nonce)
+            self.end_of_nonce_reservation = reservation_end
+            if self.enc_cipher.iv != PREFIX + long_to_bytes(next_nonce):
```
do you really need this read-back of the iv?

textshell replied:
Yes, we can just unconditionally reset the cipher, should no harm anything.

ThomasWaldmann commented on src/borg/key.py:
```diff
@@ -154,13 +156,82 @@ class AESKeyBase(KeyBase):
+            if self.enc_cipher.iv != PREFIX + long_to_bytes(next_nonce):
+                self.enc_cipher.reset(self.enc_key, PREFIX + long_to_bytes(next_nonce))
+        else:
+            next_nonce = bytes_to_long(self.enc_cipher.iv[8:])
```
hmm, here it is also reading back the iv.

i am asking because reading back the iv was a bit dirty in openssl 1.0 and does not work any more in 1.1.
also, it doesn't work for gcm, iirc.

ThomasWaldmann commented on src/borg/key.py:
```diff
@@ -154,13 +156,82 @@ class AESKeyBase(KeyBase):
+                    self.commit_local_nonce_reservation(reservation_end, local_free_nonce)
+                    self.end_of_nonce_reservation = reservation_end
+                    self.enc_cipher.reset(self.enc_key, PREFIX + long_to_bytes(next_nonce))
+
```
somehow thinking whether this whole nonce mgmt should go into a NonceManager class.

ThomasWaldmann commented on src/borg/repository.py:
```diff
@@ -173,6 +173,21 @@ def load_key(self):
 
+    def get_free_nonce(self):
+        nonce_path = os.path.join(self.path, 'nonce')
+        if os.path.exists(nonce_path):
```
how about instead exists() + open() just catching the exception if it does not open, because file is missing?

textshell replied:
Yes, that might be cleaner in python.

ThomasWaldmann commented on src/borg/key.py:
```diff
@@ -18,14 +18,20 @@
 
 
 PREFIX = b'\0' * 8
+NONCE_SPACE_RESERVATION = 1000
```
is that for 1000 * 16 bytes data to encrypt?

guess we rather go to 2**28 or so.

textshell replied:
Yes, i think we want this bigger. But going all the way to 4GB seems to be quite much. The way this currently works is that even a delete will consume 4GB of nonce space. Not sure that’s a problem though. 2**64 is a big number. Still gives 2**32 deletes till the nonce space is exhausted.

ThomasWaldmann commented on src/borg/key.py:
```diff
@@ -137,6 +143,75 @@ def decrypt(self, id, data):
+            self.end_of_nonce_reservation = reservation_end
+            self.enc_cipher.reset(None, PREFIX + long_to_bytes(next_nonce))
+        else:
+            next_nonce = bytes_to_long(self.enc_cipher.iv[8:])
```
why throw away 8 bytes?

textshell replied:
Yes, in that case it would also be easier to assert that the nonce didn‘t grow to large for the chunk format to encode. I copied the usage from other parts where they did make more sense.

But i like to keep the external storage formated as 8 bytes, because that is the usable space for the legacy encryption.

ThomasWaldmann commented on src/borg/key.py:
```diff
@@ -137,6 +143,75 @@ def decrypt(self, id, data):
+                    self.commit_repo_nonce_reservation(reservation_end, repo_free_nonce)
+                    self.commit_local_nonce_reservation(reservation_end, local_free_nonce)
+                    self.end_of_nonce_reservation = reservation_end
+                    self.enc_cipher.reset(None, PREFIX + long_to_bytes(next_nonce))
```
see above. we could just do the computations with 16bytes.

enkore commented:
> Depends on encore/issue/1060 for SaveFile

testing that right now, if I don't find any issues and CR-wise no problems we can just merge it with this.

ThomasWaldmann commented:
travis is a bit unhappy currently.

textshell commented:
I hope the existing tests run now. I still have to add new ones.

ThomasWaldmann commented on src/borg/key.py:
```diff
@@ -154,13 +226,17 @@ class AESKeyBase(KeyBase):
     PAYLOAD_OVERHEAD = 1 + 32 + 8  # TYPE + HMAC + NONCE
 
+    def __init__(self, repository):
+        super(AESKeyBase, self).__init__(repository)
```
looks like this is not needed any more.

btw, python3 has argumentless super() for the usual case.

textshell replied:
Right, there was a better syntax. I’ll fix that.

ThomasWaldmann commented on src/borg/remote.py:
```diff
@@ -56,6 +56,8 @@ class RepositoryServer:  # pragma: no cover
         'load_key',
         'break_lock',
+        'get_free_nonce',
+        'commit_nonce_reservation'
```
hmm, just an idea:

could this be also stored into the manifest (and loaded / saved with the manifest) instead of adding new api functions to the repo?

textshell replied:
No, that won’t work. We need to be able to commit the reservation independently of normal repository commiting. And we need to be able to save the reservation without using the main encryption key.
So don’t think there is any way to include it into the normal chunk storage. And i don‘t think writing it into the config file would be any better (still needs new methods, it really doesn’t belong there). The semantics just don’t match with what we currently have.

ThomasWaldmann replied:
ok

textshell commented:
I finally got around to rebase to current master and finish tests for this change. This also includes two fixes found while testing.

While the test cases are not as neat as i would like, they should have a fairly good coverage both line wise as well as testing the relevant situations where this protection is needed.

Still open is what kind of locking is needed on the client side. Usually the cache should be open when data is encrypted so that could be ok as lock. As far as i can see all of the usual commands that do encryption have an open cache. Maybe some one time conversion methods don‘t.

We need to finally decide if the iv space reservation size currently used is ok. It‘s quite big so it should be very well performance wise. On the other side a simple delete consumes IV space that otherwise could have been used to store 4GB data to the repository. But still the IV space is good for 68G small operations (i.e. writing less than 4GB), so no real worries there.

I originally thought it would be good to have something to check if a repository had any nonce reusage problems. But i don‘t that is really needed. The realistic case where reuse is possible is with rollbacks, which can‘t be detected from the repository state later anyway.

ThomasWaldmann commented on src/borg/repository.py:
```diff
@@ -190,6 +190,9 @@ def load_key(self):
 
     def get_free_nonce(self):
+        if not self.lock.got_exclusive_lock():
+            raise AssertionError("bug in code, exclusive lock should exist here")
```
hmm, why not: assert got_lock, "bug ..."

textshell replied:
I copied from prepare_txn, but yes, here the situation is simpler so i’ll just use a normal assert. Do we worry about python removing these checks in optimize mode? That would be the only reason to keep it as explicit raise.

enkore replied:
raise is the proper way to go here imho, it's not a casual "oh btw." assert, but something we wouldn't want to lose just because someone (user, packager, admin, ...) thought that -O globally is a good idea to improve performance by 0.1 % ;D

ThomasWaldmann replied:
ok, keep it.

ThomasWaldmann commented on src/borg/key.py:
```diff
@@ -197,7 +197,7 @@ def ensure_reservation(self, nonce_space_needed):
         else:
             next_nonce = int.from_bytes(self.enc_cipher.iv, byteorder='big')
             assert(next_nonce <= self.end_of_nonce_reservation)
```
assert is not a function call, but it looks like one here.

enkore commented:
Re. reservation size: it could maybe be smaller, but it _doesn't really matter_ the way you implemented it, since either you'll encrypt lots of data, then it doesn't matter, since the nonce space will be actually used, or you don't meaning some maintenance op like delete or prune is done, where normally you'd expect less than one reservation used, so yeah, billions of ops and/or tons of data encrypted. So, I think this is ok.

enkore commented on src/borg/testsuite/key.py:
```diff
@@ -77,13 +94,23 @@ def test_keyfile(self, monkeypatch, keys_dir):
+        monkeypatch.setenv('BORG_PASSPHRASE', 'test')
+        repository = self.MockRepository()
+        with open(os.path.join(get_key_nonces_dir(), repository.id_str), "w") as fd:
+            assert fd.write("0000000000002000")
```
not really meaningful assert

textshell replied:
i’ll remove that.

ThomasWaldmann commented on src/borg/testsuite/key.py:
```diff
@@ -232,3 +248,225 @@ def test_passphrase_new_retries(self, monkeypatch):
+        def commit_nonce_reservation(self, next_unreserved, start_nonce):
+            assert start_nonce == self.next_free
+            self.next_free = next_unreserved
+            pass
```
remove the pass?

ThomasWaldmann commented on src/borg/testsuite/key.py:
```diff
@@ -232,3 +248,225 @@ def test_passphrase_new_retries(self, monkeypatch):
+        assert self.repository.next_free == 0x2034
+
+    def test_transaction_abort_old_server(self, monkeypatch):
+        monkeypatch.setattr(borg.key, 'NONCE_SPACE_RESERVATION', 32)
```
the hex / dec mix here is a bit unfortunate. esp. if you use 32 dec and 20 dec, it might confuse more than with other values.

textshell replied:
i can use hex everywhere. Most of the values are readable as hex too. Maybe 20 (dec) was not the greatest choice. But 0x14 isn‘t too bad.

ThomasWaldmann replied:
i don't mean having to convert 20 to 0x14.
if mean using 32 dec (== 20 hex) and 20 dec.

ThomasWaldmann commented on src/borg/testsuite/repository.py:
```diff
@@ -390,6 +390,48 @@ def test_additional_free_space(self):
+            with pytest.raises(Exception):
+                self.repository.commit_nonce_reservation(200, 15)
+
+            self.repository.commit_nonce_reservation(512, 200)
```
similar here: 512 dec and 200 dec.

enkore commented on src/borg/key.py:
```diff
@@ -147,6 +150,75 @@ def decrypt(self, id, data, decompress=True):
+            self.commit_local_nonce_reservation(reservation_end, local_free_nonce)
+            self.end_of_nonce_reservation = reservation_end
+            self.enc_cipher.reset(None, next_nonce.to_bytes(16, byteorder='big'))
+        else:
```
"if not ...: ... else:" <- un-invert the if and swap the two blocks?

textshell replied:
this order puts initialization before the normal code. But i’m happy to swap it around if you think it does make it more readable.

enkore replied:
ah, indeed. in that case it's probably better this way around

enkore commented on src/borg/key.py:
```diff
@@ -147,6 +150,75 @@ def decrypt(self, id, data, decompress=True):
+                    self.commit_repo_nonce_reservation(reservation_end, repo_free_nonce)
+                    self.commit_local_nonce_reservation(reservation_end, local_free_nonce)
+                    self.end_of_nonce_reservation = reservation_end
+                    self.enc_cipher.reset(None, free_nonce_space.to_bytes(16, byteorder='big'))
```
it looks to me like the three blocks above share quite a bit of logic and code, try to deduplicate that?

re = .., assert, self.crnr, self.clnr, self.eonr look identical, some differences beyond that...

textshell replied:
with maximum deduplication it looks like this. It‘s quite a bit reordered, fast path up front.
I’m not quite sure if i like it like this. But maybe even if it‘s not easier to understand it might be easier to maintain.
```
    def ensure_reservation(self, nonce_space_needed):
        if self.end_of_nonce_reservation:
            # we already got a reservation, if nonce_space_needed still fits everything is ok
            next_nonce = int.from_bytes(self.enc_cipher.iv, byteorder='big')
            assert next_nonce <= self.end_of_nonce_reservation
            if next_nonce + nonce_space_needed <= self.end_of_nonce_reservation:
                return

        repo_free_nonce = self.get_repo_free_nonce()
        local_free_nonce = self.get_local_free_nonce()
        free_nonce_space = max(x for x in (repo_free_nonce, local_free_nonce, self.manifest_nonce, self.end_of_nonce_reservation) if x is not None)
        reservation_end = free_nonce_space + nonce_space_needed + NONCE_SPACE_RESERVATION
        assert reservation_end < MAX_REPRESENTABLE_NONCE
        if self.end_of_nonce_reservation is None:  # init
            self.enc_cipher.reset(None, free_nonce_space.to_bytes(16, byteorder='big'))
        else:
            # try to expand reservation
            if free_nonce_space != self.end_of_nonce_reservation:
                # some other client got an interleaved reservation, skip partial space in old reservation to avoid overlap
                self.enc_cipher.reset(None, free_nonce_space.to_bytes(16, byteorder='big'))
        self.commit_repo_nonce_reservation(reservation_end, repo_free_nonce)
        self.commit_local_nonce_reservation(reservation_end, local_free_nonce)
        self.end_of_nonce_reservation = reservation_end
```

textshell replied:
actually if the comments about init and expanding and interleaved are considered to be unimportant it could even be for the whole conditional following `if self.end_of_nonce_reservation is None`
```python
        if free_nonce_space != self.end_of_nonce_reservation:
            self.enc_cipher.reset(None, free_nonce_space.to_bytes(16, byteorder='big'))
```
But i think this it would quite a bit less meaningful. Maybe a big comment could help there, but...

enkore commented on src/borg/testsuite/key.py:
```diff
@@ -9,7 +10,12 @@
+from ..key import NonceManager
+from ..remote import InvalidRPCMethod
+
+import borg.key  # for monkey patching NONCE_SPACE_RESERVATION
```
from .. import key?

textshell replied:
For some reason i had problems with relative imports. But your suggesting works.
We could use

    from .. import key as key_module

Importing it as key would be quite confusing in the tests.

ThomasWaldmann commented on src/borg/helpers.py:
```diff
@@ -269,6 +269,17 @@ def get_keys_dir():
+    """Determine where to store the local nonce high watermark"""
+
+    xdg_config = os.environ.get('XDG_CONFIG_HOME', os.path.join(get_home_dir(), '.config'))
+    nonces_dir = os.environ.get('BORG_KEY_NONCES_DIR', os.path.join(xdg_config, 'borg', 'key-nonces'))
```
hmm, why "key" nonces? just nonces?

textshell replied:
My thinking was it’s where the nonces for the keyfiles/repokeys are stored. Also key does have the right associations for people who do not even know what a nonce is.

ThomasWaldmann replied:
but the nonces are used to encrypt repo data, not to encrypt the key.

textshell replied:
Sure. It was meant as the nonce being a mutable property of the key and thus stored at distinct location but still logically related to the key.
If you prefer BORG_NONCES_DIR i’ll use that.

enkore replied:
KEY_NONCES_DIR is good i think

ThomasWaldmann commented on src/borg/key.py:
```diff
@@ -147,6 +150,75 @@ def decrypt(self, id, data, decompress=True):
+        self.repository.commit_nonce_reservation(next_unreserved, start_nonce)
+
+    def ensure_reservation(self, nonce_space_needed):
+        if not self.end_of_nonce_reservation:
```
... is None:  # init

textshell replied:
Yes, that‘t more to the point than the not. If we don‘t go with the one of the fast path first variants i will do this.

ThomasWaldmann commented on src/borg/key.py:
```diff
@@ -147,6 +150,75 @@ def decrypt(self, id, data, decompress=True):
+                    self.commit_local_nonce_reservation(reservation_end, local_free_nonce)
+                    self.end_of_nonce_reservation = reservation_end
+                else:
+                    # need an entirly new reservation
```
entirely?

textshell replied:
yes.

ThomasWaldmann commented:
move the nonce code / tests to own modules?

textshell commented:
I'll move them to a noncemanager.py module.

ThomasWaldmann commented:
or just nonces.py?

ThomasWaldmann commented:
`<nitpick>"its own module" </nitpick>`

ThomasWaldmann commented:
IMHO:
- don't merge into 1.0-maint - it is a rather big change
- merge into master now, so it gets into 1.1.0b1, collapse changesets before merging

textshell commented:
ok, i‘ve collapsed the changes into one commit and updated the commit message.
I also added a message when used with an older server because the remote code prints to stderr. When using the a new client with a old server it look like this now:

    borg delete ssh://martin@localhost/tmp/borg-test2::a15 
    Remote: RPC method get_free_nonce is not valid
    Please upgrade to borg version 1.1+ on the server for safer AES-CTR nonce handling.

enkore commented on src/borg/nonces.py:
```diff
@@ -0,0 +1,87 @@
+import os, sys
```
pep8: split in two lines

enkore commented on src/borg/nonces.py:
```diff
@@ -0,0 +1,87 @@
+        self.commit_repo_nonce_reservation(reservation_end, repo_free_nonce)
+        self.commit_local_nonce_reservation(reservation_end, local_free_nonce)
+        self.end_of_nonce_reservation = reservation_end
+
```
pep8: blank line too much

enkore commented:
travis has two remarks (^), then this can be merged

textshell commented:
I hope this is fixed. My local flake8 check didn’t complain.

enkore commented:
Merged! Thanks! :)

textshell commented:
Thanks very much for helping along the way.
